plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php

```diff
@@ -1907,6 +1946,55 @@ public function delete( &$order, $args = array() ) {
 		}
 	}
 
+	/**
+	 * Handles the deletion of an order from the orders table when sync is disabled:
+	 *
+	 * If the corresponding row in the posts table is of placeholder type,
+	 * it's just deleted; otherwise a "deleted_from" record is created in the meta table
+	 * and the sync process will detect these and take care of deleting the appropriate post records.
+	 *
+	 * @param int $order_id Th id of the order that has been deleted from the orders table.
+	 * @return void
+	 */
+	protected function handle_order_deletion_with_sync_disabled( $order_id ): void {
+		global $wpdb;
+
+		$post_type = $wpdb->get_var(
+			$wpdb->prepare( "SELECT post_type FROM {$wpdb->posts} WHERE ID=%d", $order_id )
+		);
+
+		if ( DataSynchronizer::PLACEHOLDER_ORDER_POST_TYPE === $post_type ) {
+			$wpdb->query(
+				$wpdb->prepare(
+					"DELETE FROM {$wpdb->posts} WHERE ID=%d OR post_parent=%d",
+					$order_id,
+					$order_id
+				)
+			);
+		} else {
+			$related_order_ids   = $wpdb->get_col(
+				$wpdb->prepare(
+					"SELECT id FROM {$wpdb->posts} WHERE post_parent = %d",
+					$order_id
+				)
+			);
+			$related_order_ids[] = $order_id;
+
+			// phpcs:disable WordPress.DB.SlowDBQuery
+			foreach ( $related_order_ids as $id ) {
```

Comment:
picking it up in the select query when doing the sync using a where clause on `parent` would probably be faster then inserting for all children here, wdyt?

Reply:
Sorry, I'm not following you here. The deletion records need to be created precisely when sync is disabled, right? So what do you mean by "picking it up in the select query when doing the sync"?

Reply:
So I mean that when [we select the orders that we want to delete](https://github.com/woocommerce/woocommerce/pull/37050/files/740e5e55a82708f3d04a6211f3ec44f675f3fa62#diff-1e070cc170d3b65cd6ba61b60cbd0697c8431fda674b4720926c8fbc1afecce7R365) here, we can add a clause also to pick children of those orders. That way we won't have to insert extra records during the deletion process (potentially making it slow). Since sync happens in the context of an action and not a request, a potentially slow query there is more acceptable.

Reply:
I think I get what you mean. So:

1. Don't create deletion records for child orders that need deletion.
2. Instead, at the actual deletion time, pick the child orders of those that have a deletion record and also delete them.

I understand the reasoning about the performance gain, but I don't like this approach because:

1. It makes things more difficult to understand. You reasoned that these deletion records are an extra moving part to keep track of; if additionally we change the logic to "child orders don't get a record but need to be deleted too" it's even more intricate. (It also makes manual debugging more difficult: you can't get a glimpse of how many orders are pending deletion by just counting the deletion records anymore)
2. It makes the code more complicated. `get_deleted_order_ids` has a `$limit` argument, how do we handle this when the limit includes only some of the childs of a given order? Also the code that counts pending deletions would need changes too, it's not just "count of deletion records" anymore.
3. I don't think the performance difference would be so big. How often do orders have childs and how many? These are usually refund and an order will have at most a few of these, right?

Reply:
Fair enough, one last request would be if we can insert them in bulk instead of looping over them and firing multiple queries?

Reply:
Ok, I think I can do that.

Reply:
Now I think I'm overcomplicating things. So two things can happen when deleting an order:

1. Upshifting is active for the order type. Then child orders will have their parent ids changed and thus no deletion records are needed for them.
2. Upshifting is *not* active for the order type. Then `upshift_or_delete_child_orders` will have deleted all the child orders, and this will have caused all the required deletion records for them to be created.

Thus, this `foreach` could be removed altogether, as there will be no remaining child orders at this point (all are either no longer children of the order, or they have been deleted and got their deletion record already). Does that make sense?

Reply:
Ok, change implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/49caf09db7b7b2c9c06f84846350e1905a51cfd4